src/librustc_mir/shim.rs

```diff
@@ -98,14 +99,25 @@ fn make_shim<'a, 'tcx>(tcx: ty::TyCtxt<'a, 'tcx, 'tcx>,
         ty::InstanceDef::DropGlue(def_id, ty) => {
             build_drop_shim(tcx, def_id, ty)
         }
+        ty::InstanceDef::CloneShim(def_id, ty) => {
+            let name = tcx.item_name(def_id).as_str();
+            if name == "clone" {
+                build_clone_shim(tcx, def_id, ty)
+            } else if name == "clone_from" {
+                debug!("make_shim({:?}: using default trait implementation", instance);
+                return tcx.optimized_mir(def_id);
```

Comment:
Don't you want to make `clone_from` call `clone_from` like the derived impl? Of course it would be the best to unsize and call `clone_from_slice` for arrays - maybe make *that* a lang-item?